### PR TITLE
Update bokeh viewers to version 2.0, remove bokeh viewer tool and WaveformDisplay for now

### PR DESCRIPTION
**Pull request by maxnoe**
This is to address several issues at once, it is basically a rewrite of the bokeh based visualization tools.

This should in the end fix

* #1330 
* #1247 
* #1490 

* [X] CameraDisplay
* [x] ArrayDisplay

Removed:
* WaveFormDisplay
* Tool

### CODE REVIEW

**Comment by kosack** [comment]
As you re-write the Bokeh tools, will you support using them in Jupyter Notebooks? I think that would be a good feature (mentioned in #1247).  The old version did not make this easy to do.  Basically just either exposing the handle needed to run `push_notebook(handle)` or hiding that internally would be nice. Making a quick interactive display in a notebook is a nice use case, and with the current implementation you can't do it with either the MPL or Bokeh version.  the MPL version is quite interactive when you run it in a window, but it seems most people do not do that.

**Comment by kosack** [comment]
I recall one question came up when I was thinking about this before: the current API for CameraDisplay assumes a single CameraGeometry is used to construct it.  That was mainly because it was so slow to construct a CameraDisplay (due to limitations of matplotlib), that it was impossible to have it switch geometries between draws without the speed being terrible.   The matplotlib versions are quite fast once constructed

For Bokeh, this isn't a problem, so we might consider having a special API for the Bokeh version that allows both the geometry and image to be updated.

**Comment by maxnoe** [comment]
@kosack I started from your proposal in the issue you linked, it contains the code for the notebook.

> I recall one question came up when I was thinking about this before: the current API for CameraDisplay assumes a single CameraGeometry is used to construct it.

I will see how easy that is

**Comment by maxnoe** [comment]
This works nicely in a notebook now:

Screenshot_20201028_091232


And you can also run animations using `bokeh serve` with multiple camera images:
```
"""
Run with bokeh server examples/bokeh_cameras.py
"""
import astropy.units as u
from ctapipe.visualization.bokeh import CameraDisplay
from ctapipe.instrument import CameraDescription, PixelShape, SubarrayDescription, TelescopeDescription, OpticsDescription
from ctapipe.image.toymodel import Gaussian
from bokeh.layouts import row
from bokeh.io import curdoc
import numpy as np
from ctapipe.io.toymodel import ToyEventSource



whipple_optics = OpticsDescription('Whipple', equivalent_focal_length=14.6 * u.m, num_mirrors=1, num_mirror_tiles=248)

telescopes = {
    1: TelescopeDescription.from_name('LST', 'LSTCam'),
    2: TelescopeDescription.from_name('SST-ASTRI', 'CHEC'),
    3: TelescopeDescription('Whipple', 'MST', optics=whipple_optics, camera=CameraDescription.from_name('Whipple109')),
}

subarray = SubarrayDescription(
    'bokeh_example',
    tel_positions={1: [0, 0, 0] * u.m, 2: [50, 0, 0] * u.m, 3: [0, 50, 0] * u.m},
    tel_descriptions=telescopes,
)

subarray.tel[3].camera.geometry.pix_type = PixelShape.CIRCLE

displays = {
    tel_id: CameraDisplay(tel.camera.geometry, autoshow=False)
    for tel_id, tel in subarray.tel.items()
}

for d in displays.values():
    d.add_colorbar()


source = ToyEventSource(
    subarray, max_events=100, trigger_probability=1,
    min_length_m=[
        ('id', 1, 0.1),
        ('id', 2, 0.01),
        ('id', 3, 0.01),
    ],
    max_length_m=[
        ('id', 1, 0.5),
        ('id', 2, 0.05),
        ('id', 3, 0.05),
    ],
)

event_iter = iter(source)


def update():

    event = next(event_iter)

    for tel_id, disp in displays.items():
        image = event.dl1.tel[tel_id].image

        disp.clear_overlays()
        disp.image = image
        # h = HillasParametersContainer(x=center[0], y=center[1], length=length, width=width, psi=psi)
        # disp.overlay_moments(h, line_color='green', line_width=3)

update()
curdoc().add_root(row(*[d.figure for d in displays.values()]))
curdoc().add_periodic_callback(update, 2000)
```

Screenshot_20201028_092955

**Comment by maxnoe** [comment]
With that I unfortunately found out, that the toy event source does not really generate nice images for smaller cameras.

**Comment by maxnoe** [comment]
>  the MPL version is quite interactive when you run it in a window, but it seems most people do not do that.

Did you try using the `%matplotlib notebook` backend?

**Comment by maxnoe** [comment]
Added a first version of the ArrayDisplay, now support yet for showing values or hillas vectors:

Screenshot_20210416_141021

**Comment by LukasNickel** [comment]
Also a high number of codacy complaints, have you checked them?

File: ctapipe/io/toymodel.py [hunk 1]
```diff
@@ -129,7 +129,10 @@ def generate_event(self):
             width = self.calc_width(eccentricity, length)
 
             psi = self.rng.uniform(0, 360)
-            intensity = self.rng.poisson(int(1e5 * width * length))
```
[the remark this answers is not in the document]
Comment (maxnoe) [reply ctapipe/io/toymodel.py:132]: This is the number of pixels "hit" by the shower, not the number of pixels of the camera.

**Comment by maxnoe** [comment]
I think I fixed most of the valid codacy complains

**Review by kosack — APPROVED** [review]

**Review by LukasNickel — DISMISSED** [review]
We should probably note that the python version in the enviroment changed

**Review by LukasNickel — APPROVED** [review]

**Review by kosack — COMMENTED** [review]
Testing this, I tried the following:
```py3
im = event.dl1.tel[13].image
geom = source.subarray.tel[13].camera.geometry
disp = CameraDisplay(geom, image=im)
```
Which gives me an error:
```
ValueError: expected an instance of type Title, got FlashCam (CameraFrame) of type str
```
from the following line:
```py3
    174                 title = f"{geometry} ({frame})"
--> 175             self.figure.title = title
```

Any idea why that is a problem? This works with the matplotlib version

**Comment by maxnoe** [comment]
Strange, it definitely worked when testing. Will add a unit test and see if I can get it working.

**Comment by kosack** [comment]
@maxnoe I think it's ok. I just restarted everything and it works now... Not sure why just restarting the kernel after upgrading bokeh didn't work

**Review by kosack — DISMISSED** [review]

**Comment by kosack** [comment]
Would be nice to have an example of using this in a notebook, with interactive updating. But that can come later.  I enabled `use_notebook=True`, and it displays, but when I update the image, it doesn't update automatically (even though I see a call to `push_notebook()` in `update()`).  Maybe I'm using it incorrectly?

File: .pylintrc [hunk 2]
```diff
@@ -425,7 +425,8 @@ good-names=i,
            x,
            y,
            n,
-		   f,
+		       f,
```
Comment (LukasNickel) [inline .pylintrc:428]: Huh?

Comment (maxnoe) [reply .pylintrc:428]: tabs vs spaces

File: ctapipe/visualization/bokeh.py [hunk 3]
```diff
@@ -1,408 +1,634 @@
...
+        self,
+        geometry: CameraGeometry = None,
+        image=None,
+        cmap="inferno",
+        norm="lin",
+        autoscale=True,
+        title=None,
+        # bokeh specific options
+        use_notebook=None,
+    ):
```
Comment (kosack) [inline ctapipe/visualization/bokeh.py:143]: missing docstring here. In particular, use_notebook needs to be defined as it is not in the MPL version and None has a special meaning (to auto-detect)

**Review by kosack — DISMISSED** [review]

**Comment by maxnoe** [comment]
CI error fixed here: #1858

**Review by kosack — APPROVED** [review]

**Comment by maxnoe** [comment]
This should finally build now

File: setup.py [hunk 4]
```diff
@@ -57,7 +56,7 @@
         "requests",
         "setuptools_scm>=3.4",
         "importlib_resources;python_version<'3.9'",
-        "jinja2~=3.0.2",  # for bokeh=1, remove when updaiting bokeh
+        "jinja2~=3.0.2",  # for bokeh sphinx 3.5, update when moving to 4.x
```
Comment (LukasNickel) [inline setup.py:59]: That bokeh does not belong here, right?

Comment (maxnoe) [reply setup.py:59]: fixed

**Review by LukasNickel — APPROVED** [review]

**Review by kosack — APPROVED** [review]

**Pull request merged**